router/router.go

```diff
@@ -141,8 +141,14 @@ func (r *routeSet) ProcessRoute(uri *url.URL) (newurl string, requirePermission
 }
 
 func (r *routeSet) GetProxy(req *ms_http.Request) (*url.URL, error) {
-	route := r.match(req.URL)
-	if route == nil || !route.target.useProxy {
+	var newRoute *route
```

Comment:
I think we should keep the performance characteristics consistent. More specifically, route matching as performed by routeSet.match is, loosely speaking, O(1). This however, is O(n) on the number of routes.
We should have the same performance characteristics here as we have for routeSet.match, because GetProxy is called for every single request, so it's nonsensical to go to all the effort to make routeSet.match O(1), but then leave GetProxy O(n).

I don't think we need as sophisticated matching for the targets as we have for the sources (ie the cascaded hash tables, one for each unique prefix length). Instead, I think it's perfectly fine to just keep a single hash table (aka Go map) which has host names as keys, and *route as values. You would build up this map at initialization time, and it should be a configuration error if "UseProxy" is inconsistently specified twice for the same target host.